tests/performance/local/condition_variable_overhead.cpp

```diff
@@ -0,0 +1,247 @@
+    {
+        // -------------------
+        // PING
+        // -------------------
+        // sleep until thread B notifies us
+        {
+            std::unique_lock lk1(tda->mtx_);
+            tda->cv_.wait(lk1, [tda] { return tda->ping_ready_; });
+        }
+        DEBUG("A : step 1 complete");
+
+        // notify thread B that we are ready
+        {
+            std::lock_guard lk1(tdb->mtx_);
+            tdb->pong_ready_ = false;
+            tdb->ping_ready_ = true;
+        }
+        tdb->cv_.notify_one();
+        DEBUG("A : step 2 complete");
+
+        // -------------------
+        // PONG
+        // -------------------
+        // A sleeps until thread B sets pong_ready and wakes us
+        {
+            std::unique_lock lk2(tda->mtx_);
+            tda->cv_.wait(lk2, [tda] { return tda->pong_ready_; });
+        }
+        DEBUG("A : step 3 complete");
+
+        // now thread B will go to sleep, so wake it up
+        {
+            std::lock_guard lk2(tdb->mtx_);
+            tdb->ping_ready_ = false;
+            tdb->pong_ready_ = true;
+        }
+        // wake up thread B
+        tdb->cv_.notify_one();
+        DEBUG("A : step 4 complete");
+    }
+    //
+    task_counter--;
+    DEBUG("A complete " << task_counter);
+}
+
+// ------------------------------------------------
+void function_B(std::uint64_t loops, itp tda, itp tdb)
+{
+    for (std::uint64_t i = 0; i < loops; ++i)
+    {
+        // -------------------
+        // PING
+        // -------------------
+        // notify thread A that we are ready
+        {
+            std::lock_guard lk1(tda->mtx_);
+            tda->pong_ready_ = false;
+            tda->ping_ready_ = true;
+        }
+        tda->cv_.notify_one();
+        DEBUG("B : step 1 complete");
+
+        // sleep until thread A wakes us up
+        {
+            std::unique_lock lk1(tdb->mtx_);
+            tdb->cv_.wait(lk1, [tdb] { return tdb->ping_ready_; });
+        }
+        DEBUG("B : step 2 complete");
+
+        // -------------------
+        // PONG
+        // -------------------
+        // notify thread A that we are ready
+        {
+            std::lock_guard lk2(tda->mtx_);
+            tda->ping_ready_ = false;
+            tda->pong_ready_ = true;
+        }
+        tda->cv_.notify_one();
+        DEBUG("B : step 3 complete");
+
+        // sleep until thread A wakes us up
+        {
+            std::unique_lock lk2(tdb->mtx_);
+            tdb->cv_.wait(lk2, [tdb] { return tdb->pong_ready_; });
+        }
+        DEBUG("B : step 4 complete");
+    }
+    //
+    task_counter--;
+    DEBUG("B complete " << task_counter);
+}
+
+// ------------------------------------------------
+template <typename Scheduler>
+void test_cv(Scheduler&& sched, std::uint64_t loops)
+{
+    int N = pika::get_num_worker_threads() / 2;
+    task_counter = 2 * N;
+    for (int n = 0; n < N; n++)
+    {
+        itp tda(new task_data);
+        itp tdb(new task_data);
```

Comment:
Do these need to be reference counted if we're waiting for the tasks to complete at the end?

Reply:
I'm going to leave it as is, because looking at it again, if there are only going to be 2 tasks, then ok, but if there are more, we can't easily pass the task_data things into N * (A and B) by ref unless we keep them in scope in the main part, then we must put all tasks into a vector of senders and wait on all of them.